src/riak_kv_compile_tab.erl

```diff
@@ -25,155 +25,170 @@
--spec new_table_result({ok, dets:tab_name()} | {error, any()}) ->
-                       {ok, dets:tab_name()} | error.
-new_table_result({ok, Dets}) ->
-    %% Clean up any lingering records stuck in the compiling state
-    mark_compiling_for_retry(),
-    {ok, Dets};
-new_table_result({error, Reason}) ->
-    lager:error("Could not open ~p because of ~p", [?TABLE, Reason]),
-    error.
+%%
+-spec new(file:name()) -> {ok, dets:tab_name()} | error.
+new(Dir) ->
+    {ok, _} = dets:open_file(?TABLE2, [{file, file_v2(Dir)} | ?DETS_OPTIONS]),
+    {ok, _} = dets:open_file(?TABLE3, [{file, file_v3(Dir)}, {keypos, #row_v3.table_version } | ?DETS_OPTIONS]).
 
 %% Useful testing tool
 -spec delete_dets(file:name()) ->
     ok | {error, any()}.
 delete_dets(FileDir) ->
-    FilePath = filename:join(FileDir, [?TABLE, ".dets"]),
-    dets:close(FilePath),
-    file:delete(FilePath).
+    _ = dets:close(file_v3(FileDir)),
+    _ = file:delete(file_v3(FileDir)).
 
 %%
--spec insert(BucketType :: binary(),
-             DDLVersion :: riak_ql_component:component_version(),
-             DDL :: term(),
-             CompilerPid :: pid(),
-             State :: compiling_state()) -> ok | error.
-insert(BucketType, DDLVersion, DDL, CompilerPid, State) ->
-    lager:info("DDL DETS Update: ~p, ~p, ~p, ~p, ~p",
-               [BucketType, DDLVersion, DDL, CompilerPid, State]),
-    Result2 = dets:insert(?TABLE, {BucketType, DDLVersion, DDL, CompilerPid, State}),
-    dets:sync(?TABLE),
-    insert_result(Result2).
+file_v2(Dir) ->
+    filename:join(Dir, [?TABLE2, ".dets"]).
 
--spec insert_result(ok | {error, any()}) -> ok | error.
-insert_result(ok) ->
-    ok;
-insert_result({error, Reason}) ->
-    lager:error("Could not write to ~p because of ~p", [?TABLE, Reason]),
-    error.
-
-%% Check if the bucket type is in the compiling state.
--spec is_compiling(BucketType :: binary()) ->
-    {true, pid()} | false.
-is_compiling(BucketType) ->
-    case dets:lookup(?TABLE, BucketType) of
-        [{_,_,_,Pid,compiling}] ->
-            {true, Pid};
-        _ ->
-            false
-    end.
+%%
+file_v3(Dir) ->
+    filename:join(Dir, [?TABLE3, ".dets"]).
 
 %%
--spec get_state(BucketType :: binary()) ->
-        compiling_state() | notfound.
-get_state(BucketType) when is_binary(BucketType) ->
-    case dets:lookup(?TABLE, BucketType) of
-        [{_,_,_,_,State}] ->
-            State;
-        [] ->
-            notfound
+-spec insert(BucketType :: binary(), DDL :: term()) -> ok.
+insert(BucketType, DDL) when is_binary(BucketType), is_tuple(DDL) ->
+    lager:info("DDL DETS Update: ~p, ~p", [BucketType, DDL]),
+    DDLVersion = riak_ql_ddl:ddl_record_version(element(1, DDL)),
+    Row = #row_v3{
+         table = BucketType
+        ,table_version = {BucketType, DDLVersion}
+        ,ddl = DDL},
+    ok = dets:insert(?TABLE3, Row),
+    ok = dets:sync(?TABLE3),
+    case lists:keyfind(ddl_v1, 1, riak_ql_ddl:convert(v1, DDL)) of
```

Comment:
Yes, but the DDL record capabilitiy will also be downgraded so none of the nodes will be able to use that table.